### ISSUE TITLE
INP001 false positive in tests folder

### ISSUE THREAD

**Issue by flying-sheep** [astral-sh/ruff#6474]
The recommended Python + Pytest layout is

- `src/`
  - `mypkg/`
    - `__init__.py`
    - `subpkg/`
      - `__init__.py`
- `tests/`
  - `test_thing.py`
  - `test_more/`
    - `test_other_thing.py`

Notably, `tests` and all files under it *aren’t packages*. The [recommended import mode](https://docs.pytest.org/en/7.1.x/explanation/goodpractices.html#tests-outside-application-code) enforces that you can’t import from them.

Therefore `INP001` shouldn’t trigger when `tests` is on top level of the project.

In fact, it would be great to have both a rule that enforces the opposite and to ensure that `tests/**/test_*.py` files don’t import things from each other.

**Comment by tjkuson** [comment]
I think this could be related to #6114.

**Comment by flying-sheep** [comment]
Pytest collects and imports individual files. That’s independent from Python’s namespace packages.

It’s possible that that other issue incidentally fixes this, but this is an independent issue.

**Comment by charliermarsh** [comment]
Interesting. To what degree is this specific to pytest? On the surface it feels like per-file-ignores is the right solution here…

**Comment by flying-sheep** [comment]
It’s very specific to Pytest. I’m coming from the perspective that when people have e.g.

```toml
[tool.ruff]
select = ['INP', 'PT']
```

```
project_dir
+- src/...
+- tests/
   +- test_foo.py
   +- test_bar/
      +- __init__.py
      +- test_bar_basic.py
```

then they should optimally see something like this:

```
tests/test_bar/__init__.py:1:1: PT030 File `tests/test_bar/__init__.py` found. Pytest test directories shouldn’t contain packages.
```

Currently they will see an instruction to do the opposite of the recommended route:

```
tests/test_foo.py:1:1: INP001 File `tests/test_foo.py` is part of an implicit namespace package. Add an `__init__.py`.
```

**Comment by sanmai-NL** [comment]
See also #8690

**Comment by Artalus** [comment]
For what it's worth, it seems that even _with_ the `__init__.py` file present, `ruff` still complains about the `INP001` (unlike the original flake linter):
```
[artalus@idpad foo]$ tree --gitignore
.
├── pyproject.toml
├── README.md
├── src
│   └── foo
│       └── __init__.py
├── tests
│   └── test_foo.py
└── uv.lock

[artalus@idpad foo]$ uv pip install flake8-no-pep420 ruff
Using Python 3.13.2 environment at: /tmp/420/.venv
Resolved 6 packages in 519ms
Installed 3 packages in 52ms
 + flake8==7.2.0
 + flake8-no-pep420==2.8.0
 + ruff==0.11.10

[artalus@idpad foo]$ flake8
./tests/test_foo.py:1:1: INP001 File is part of an implicit namespace package. Add an __init__.py?

[artalus@idpad foo]$ ruff check
tests/test_foo.py:1:1: INP001 File `tests/test_foo.py` is part of an implicit namespace package. Add an `__init__.py`.
Found 1 error.

[artalus@idpad foo]$ touch tests/__init__.py

[artalus@idpad foo]$ flake8

[artalus@idpad foo]$ ruff check
tests/test_foo.py:1:1: INP001 File `tests/test_foo.py` is part of an implicit namespace package. Add an `__init__.py`.
Found 1 error.
```

We ended up disabling `INP001` for the tests dir:
```toml
# /pyproject.toml
[tool.ruff.lint.per-file-ignores]
"tests/*" = [  # /* ensures it works for all subdirectories
    "INP001",  # "implicit namespace package" is invalid for pytest
]

# OR,

# /tests/ruff.toml
# inherit all settings from the main file
extend = "../pyproject.toml"
[lint]
ignore = [
    "INP001",
]
```

**Comment by calumy** [comment]
I believe the `__init__.py` file needs to be present in the test directory to fix the `INP001` error

> ```
> [artalus@idpad foo]$ tree --gitignore
.
├── pyproject.toml
├── README.md
├── src
│   └── foo
│       └── __init__.py
├── tests
│   ├── test_foo.py
│   └── __init__.py
└── uv.lock
```

(I haven't actually tested this, but I believe it is correct as we have `__init__.py` files in our test directories and don't see these errors.)

**Comment by MichaReiser** [comment]
Special casing `tests` in INP001` if it's at the root of the project seems like a reasonable default to me.

**Comment by copdips** [comment]
in my pyproject.toml:

```toml
[tool.ruff.lint.per-file-ignores]
"tests/**/*.py" = [
    "INP001",  # implicit-namespace-package
    "S101",  # assert
]
```

**Comment by bneijt** [comment]
I just want to add to this that I had a nasty time debugging an issue after blindly following Ruff and adding an `__init__.py` to my `tests` folder. I had multiple packages in a `uv` workspace:
```
packages/a/tests/__init__.py
packages/a/tests/test_main.py
packages/b/tests/__init__.py
packages/b/tests/test_main.py
```
This would "ignore" the second `test_main.py` content, but would still print out a statement that it ran. I think the `--import-mode=importlib` required to support multiple packages ends up getting confused because both files are registered as test_main inside the tests module and the first test file will be loaded when pytest tries to load the second package, then runs the tests from the a package and reports that all tests in the b packages test_main succeeded.

Took me a while to figure that one out, so I would love for an exception for tests and I'm going to use the following for now:
```
[tool.ruff.lint.per-file-ignores]
"**/tests/**/*.py" = [
    "INP001",  # Harms multi-package test discovery https://github.com/astral-sh/ruff/issues/6474
]
```

**Comment by ntBre** [comment]
While reviewing #27985, I realized that [`unittest`](https://docs.python.org/3/library/unittest.html#test-discovery) actually does expect `__init__.py` to be present in a `tests/` directory, if I'm understanding correctly:

> Changed in version 3.11: Test discovery dropped the [namespace packages](https://docs.python.org/3/glossary.html#term-namespace-package) support. It has been broken since Python 3.7. Start directory and its subdirectories containing tests must be regular package that have `__init__.py` file.

For example:

```console
$ tree
.
└── tests
    └── test_foo.py

2 directories, 1 file
$ python3.14 -m unittest discover .

----------------------------------------------------------------------
Ran 0 tests in 0.000s

NO TESTS RAN
$ python3.14 -m unittest discover tests
.
----------------------------------------------------------------------
Ran 1 test in 0.000s

OK
```

Depending on your `unittest` invocation, it might also work, but the first version seems like it could be a common usage and discovers no tests without `__init__.py`.

So it seems to me that if we add a default special case to exclude `/tests`, we also need a configuration option to re-enable it for `unittest` users. `per-file-ignores` still seems like the best solution to me. Maybe we should just update the documentation to suggest this for pytest users?

**Comment by flying-sheep** [comment]
Yeah, as @copdips pointed out, there are a few rules that are flipped or different for pytest, I don’t know how to handle it ideally, but

- pytest files **shouldn’t** have `__init__` (this issue)
- in pytest files, `assert` **should** be used, since pytest rewrites these (both in helper functions and test functions)
- pytest test functions shouldn’t trigger “too many (positional) arguments” as pytest uses dependency injection, so they’re implicitly only ever used as “keyword arguments”

Both files and functions are by default defined by their names starting with `test_`, but that’s configurable.

**Comment by LoicRiegel** [comment]
While I agree ``tests/`` is often intentionally not a package, I found a case where ``__init__.py`` is necessary for successful test collection:

```
tests/
├── a/test_sub.py
└── b/test_sub.py
```

Like this, pytest reports an "import file mismatch" because both files are imported with the same basename, and cannot distinguish them. Adding an ``__init__.py`` to ``tests/a`` and ``tests/b`` resolves the issue by giving them unique module names. Actually, just adding one in ``tests/a`` is sufficient.
```
_________________________ ERROR collecting tests/b/test_sub.py ___________________________
import file mismatch:
imported module 'test_sub' has this __file__ attribute:
  C:\Users\...\test-pytest\tests\a\test_sub.py
which is not the same as the test file we want to collect:
  C:\Users\...\test-pytest\tests\b\test_sub.py
HINT: remove __pycache__ / .pyc files and/or use a unique basename for your test file modules
```
(removing the cache does nothing)
